CHANGELOG.md

```diff
@@ -10,12 +10,318 @@ New features:
+
+It used to be possible to match on negative litterals, such as `-1`, but this prevented parsing matches on constructors aliased to `-`. The compiler will reject matches on _bare_ negative litterals, but they can still be matched by wrapping them in parentheses.
+
+* Forbid partial data constructors exports (#3872, @kl0tl)
+
+Exporting only some of the constructors of a type meant that changes internal to a module, such that adding or removing an unexported constructor, could cause unexhaustive pattern matches errors in downstream code. Partial explicit export lists will have to be completed with the missing constructors or replaced by implicit export lists.
+
+* Print compile errors to stdout, progress messages to stderr (#3839, @JordanMartinez)
+
+The compiler used to print errors to stderr, which is especially annoying when piping its output to another program with the `--json-errors` flag. Stderr is really for diagnostics about the execution of a program, whereas its expected output ought to be printed to stdout.
+
+#### Fixes
+
+* Only include direct dependencies in the output for `purs graph` instead of their transitive closure (#3993, @colinwahl)
+
+* Fix the reversal of the qualifier of qualified operators (#3971, @rhendric)
+
+Qualified operators, for instance `Data.Array.(!)`, were interpreted with a reversed qualifier, like `Array.Data.(!)`.
+
+* Check all recursive paths in data binding groups (#3936, @natefaubion)
+
+The compiler was not catching recursive type synonyms when some recursive paths were guarded by data types or newtypes.
+
+* Desugar type operator aliases inside parens (#3935, @natefaubion)
+
+The compiler did not accept type operators inside parens in prefix position, except `(->)`.
+
+* Pin language-javascript to a specific version (#3904, @hdgarrood)
+
+Allowing the compiler to be built against various versions of `language-javascript` meant that multiple builds of the same version of the compiler could accept different syntaxes for JavaScript foreign modules, depending on how they were built.
+
+#### Improvements
+
+* Improve error message when `negate` isn't imported (#3952, @mhmdanas)
+
+This shows a specific message when using negative litterals but `Data.Ring.negate` is out of scope, similar to the messages shown when using do notation if `Control.Bind.bind` and `Control.Bind.discard` are out of scope.
+
+* Add source spans to `PartiallyAppliedSynonym` errors (#3951, @rhendric)
+
+`PartiallyAppliedSynonym` errors were usually rethrown with the appropriate source span, but not when deriving instances. This annotates those errors with the source span of the partially applied synonyms themselves, which is more robust and accurate than rethrowing the error with an approximate source span.
+
+* Allow type synonyms in instances heads and superclass constraints (#3539, #3966, #3965, @garyb, @kl0tl)
+
+This allows declarations such as
+
+```purescript
+newtype App a = App (ReaderT Env Aff a)
```

Comment:
Perhaps we could explicitly include the type synonym for `Env` because it's user-defined and to make the fact it's a type synonym apparent:

```purescript
type Env = { port :: Int }
newtype App a = App (ReaderT Env Aff a)
derive newtype instance monadAskApp :: MonadAsk Env App
```

Reply:
I agree with this, too.